### PR TITLE
ROC-4396 Update calculation of 'Determination of means'

### PR DESCRIPTION
**Pull request by tharacka**
### JIRA link

https://tools.hmcts.net/jira/browse/ROC-4396

### Change description

Changes to the models and calculation of statement of means.
The UI changes have been completed in other PRs.

### Work checklist

- [x] Unit tests added where applicable
- [ ] Route tests added for new pages
- [ ] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [ ] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [ ] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/app/common/court-calculations/courtDetermination.ts [hunk 1]
```diff
@@ -12,16 +12,25 @@ export class CourtDetermination {
                             claimantPaymentDate: Moment,
                             courtGeneratedPaymentDate: Moment): DecisionType {
 
-    if (!defendantPaymentDate || !claimantPaymentDate || !courtGeneratedPaymentDate) {
+    if (!defendantPaymentDate || !claimantPaymentDate) {
```
Comment (ckalista) [inline src/main/app/common/court-calculations/courtDetermination.ts]: As discussed, overall the logic can be simplified here. Something like this is easier to reason about:

```
import { Moment } from 'moment'

export enum DecisionType {
  CLAIMANT = 'CLAIMANT',
  DEFENDANT = 'DEFENDANT',
  COURT = 'COURT'
}

export class CourtDetermination {

  static calculateDecision (defendantPaymentDate: Moment,
                            claimantPaymentDate: Moment,
                            courtGeneratedPaymentDate: Moment): DecisionType {

    if (!defendantPaymentDate || !claimantPaymentDate) {
      throw new Error('Input should be a moment, cannot be empty')
    }

    if (claimantPaymentDate.isSameOrAfter(defendantPaymentDate)) {
      return DecisionType.CLAIMANT
    }

    if (!courtGeneratedPaymentDate) {
      return DecisionType.DEFENDANT
    }

    if (courtGeneratedPaymentDate.isBefore(claimantPaymentDate)) {
      return DecisionType.COURT
    }

    if (courtGeneratedPaymentDate.isBefore(defendantPaymentDate)) {
      return DecisionType.COURT
    }
    
    return DecisionType.DEFENDANT
  }
}
```

File: src/main/app/common/court-calculations/courtDetermination.ts [hunk 2]
```diff
@@ -12,16 +12,25 @@ export class CourtDetermination {
...
+        return DecisionType.DEFENDANT
+      }
 
-    if (claimantPaymentDate.isSameOrAfter(defendantPaymentDate) || claimantPaymentDate.isSameOrAfter(courtGeneratedPaymentDate)) {
-      return DecisionType.CLAIMANT
-    }
-    if (claimantPaymentDate.isSameOrBefore(courtGeneratedPaymentDate) && defendantPaymentDate.isSameOrBefore(courtGeneratedPaymentDate)) {
+      return DecisionType.COURT
+
+    } else {
```
Comment (ckalista) [inline src/main/app/common/court-calculations/courtDetermination.ts]: Thanks for fixing this bug 👍

File: src/main/app/common/allowances/allowance.ts [hunk 3]
```diff
@@ -1,35 +1,4 @@
...
-  weekly?: number
-  monthly?: number
-
-  constructor (item?: string, weekly?: number, monthly?: number) {
-    this.item = item
-    this.weekly = weekly
-    this.monthly = monthly
-  }
-
-  static fromObject (value?: any): AllowanceItem {
```
Comment (ckalista) [inline src/main/app/common/allowances/allowance.ts]: `fromObject` methods are typically used to reconstruct an instance from a form object. I am not quite sure if we need this and if we do we probably have a problem :-)

File: src/main/app/common/allowances/allowanceItem.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  weekly?: number
+  monthly?: number
+
+  constructor (item?: string, weekly?: number, monthly?: number) {
+    this.item = item
+    this.weekly = weekly
+    this.monthly = monthly
+  }
+
+  static fromObject (value?: any): AllowanceItem {
```
Comment (ckalista) [inline src/main/app/common/allowances/allowanceItem.ts]: Same comment as in `allowance.ts`

File: src/main/app/common/allowances/allowance.ts [hunk 5]
```diff
@@ -1,35 +1,4 @@
...
-      this.item = input.item
-      this.weekly = input.weekly
-      this.monthly = input.monthly
-    }
-    return this
-  }
-}
+import { AllowanceItem } from 'common/allowances/allowanceItem'
 
 export class Allowance {
```
Comment (ckalista) [inline src/main/app/common/allowances/allowance.ts]: I can't find a test for this class - we should probably add one.

File: src/main/app/common/allowances/allowanceItem.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+export class AllowanceItem {
```
Comment (ckalista) [inline src/main/app/common/allowances/allowanceItem.ts]: I can't find a test for this class - we should probably add one.

File: src/main/app/common/allowances/allowance.ts (hunk 5 above)
Comment (ckalista) [reply src/main/app/common/allowances/allowance.ts]: In the constructor of this class, consider using usual syntactic sugar: 

```
constructor (public personal?: AllowanceItem[],
                public dependant?: AllowanceItem[],
                public pensioner?: AllowanceItem[],
                public disability?: AllowanceItem[]) 
```

File: src/main/app/common/allowances/allowanceRepository.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import { AllowanceItem } from 'common/allowances/allowanceItem'
+import { Allowance } from 'common/allowances/allowance'
+import { join } from 'path'
+
+export enum DependantAllowanceType {
```
Comment (ckalista) [inline src/main/app/common/allowances/allowanceRepository.ts]: I would have places those types in the `allowance` module instead - what do you think?

File: src/main/app/common/allowances/allowanceRepository.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  getPensionAllowance (pensionAllowanceType: PensionAllowanceType): AllowanceItem
+
+}
+
+export class ResourceAllowanceRepository implements AllowanceRepository {
+
+  private allowances: Allowance
+
+  constructor (public option?: string) {
+    let meansAllowance = option
```
Comment (ckalista) [inline src/main/app/common/allowances/allowanceRepository.ts]: Just suggesting a slightly different construct:

```
  constructor (path?: string) {
    const resourcePath = path || join(__dirname, '..', '..', '..','resources','meansAllowance.json')
    this.allowances = new Allowance().deserialize(require(resourcePath))
  }
```

File: src/main/app/common/allowances/allowanceHelper.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+export interface AllowanceHelper {
+  getMonthlyPensionerAllowance (income: Income[], partner: Partner)
+  getMonthlyDependantsAllowance (dependants: Dependant)
+  getMonthlyLivingAllowance (defendantAge: number, partner: Partner)
+  getCarerDisableDependantAmount (dependant: Dependant, isCarer: boolean)
+  getDisabilityAllowance (defendantDisability: DisabilityStatus, partner: Partner)
+}
+
+export class ResourceAllowanceHelper implements AllowanceHelper {
```
Comment (ckalista) [inline src/main/app/common/allowances/allowanceHelper.ts]: Maybe rename this `AllowanceCalculations` and get rid of the interface. Then I think it is OK to have your `StatementOfMeansCalculations` tightly coupled with `AllowanceCalculations` now that we have `AllowanceRepository` outside of `StatementOfMeansCalculations`.

**Comment by markry-kainos** [comment]
actually its not :) - but im not sure why they are coming through as changes here

**Pull request closed without merging**